specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/preview/2024-12-09-preview/confidentialledger.json

```diff
@@ -1746,6 +1917,91 @@
         "roleActions"
       ],
       "type": "string"
+    },
+    "UserDefinedFunction": {
+      "description": "A user defined function in the ledger.",
+      "required": [
+        "name"
+      ],
+      "properties": {
+        "body": {
+          "type": "string",
+          "description": "Body of the user defined function in JavaScript code."
+        },
+        "name": {
+          "type": "string",
+          "description": "Name of the user defined function."
+        }
+      },
+      "type": "object"
+    },
+    "UserDefinedFunctionExecutionProperties": {
+      "description": "The properties for executing a user defined function.",
+      "properties": {
+        "arguments": {
+          "type": "string",
+          "description": "Runtime arguments of the user defined function."
+        },
+        "runtimeOptions": {
+          "$ref": "#/definitions/JSRuntimeOptions"
+        }
+      },
+      "type": "object"
+    },
+    "UserDefinedFunctionExecutionStatus": {
+      "type": "string",
+      "description": "Represents the status of a user defined function execution.",
+      "enum": [
+        "Succeeded",
+        "Failed"
+      ],
+      "x-ms-enum": {
+        "name": "UserDefinedFunctionExecutionStatus",
+        "modelAsString": true
+      }
+    },
+    "UserDefinedFunctionExecutionResult": {
+      "description": "The result of executing a user defined function.",
+      "required": [
+        "status"
+      ],
+      "properties": {
+        "error": {
+          "type": "string",
+          "description": "Error returned when executing the user defined function in case of failure."
+        },
+        "result": {
+          "type": "string",
+          "description": "Return value of the user defined function in case of success."
+        },
+        "status": {
```

Comment:
Being able to "successfully fail" is a tricky design point. The caller is going to have to check the HTTP status and then the status property in the response body before they can check the status or result properties.

From a description perspective, we *could* consider using a discriminated type here (with the status as the discriminator field).

Reply:
I was following the example of the [status monitor resource](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#the-status-monitor-resource) to report the result of a function execution. 

Just to make sure I understand your comment properly, are you suggesting using only the `status` property to discriminate between a function execution success or failure? That's already the case because this operation already returns a 200 status code regardless if the function succeeded or failed. So, users will have to only inspect the `status` property in the response payload to know whether their function succeeded or returned an exception.

Please let me know if I'm misunderstanding or missing something else.

Reply:
I will add more details in the descriptions.

Reply:
Added